### ISSUE TITLE
Error opening local file `/dev/fd/1` & quoting issue

### ISSUE THREAD

**Issue by tukusejssirs** [sindresorhus/execa#498]
I am dealing with eflexsystems/node-samba-client#50.

Running the following snippet results in an error.

```js
const args = [
  '-U',
  'server',
  '-c',
  'get "./file.txt" "/dev/fd/1"',
  '//1.2.3.4/share',
  '--password',
  'p4ss'
]

const options = {all: true, cwd: ''}

try {
  const {all} = await execa('smbclient', args, options)
  return all
} catch (error) {
  error.message = error.all
  error.shortMessage = error.all
  throw error
}

// Error output
// Note: I am running modified version of `samba-client` in order to test this out,
//       therefore the lines numbers don't match with the official version.
Uncaught Error: Error opening local file /dev/fd/1
    at makeError ($repo_root/node_modules/samba-client/node_modules/execa/lib/error.js:60:11)
    at handlePromise ($repo_root/node_modules/samba-client/node_modules/execa/index.js:118:26)
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
    at async SambaClient.execute ($repo_root/node_modules/samba-client/index.js:184:18)
    at async SambaClient.readFile ($repo_root/node_modules/samba-client/index.js:43:10) {
  shortMessage: 'Error opening local file /dev/fd/1',
  command: 'smbclient -U server -c get "./file.txt" "/dev/fd/1" //1.2.3.4/share --password p4ss',
  escapedCommand: 'smbclient -U server -c "get \\"./file.txt\\" \\"/dev/fd/1\\"" "//1.2.3.4/share" --password p4ss',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: 'Error opening local file /dev/fd/1',
  stderr: '',
  all: 'Error opening local file /dev/fd/1',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false
}
```

`escapedCommand` contains `'smbclient -U server -c "get \\"./file.txt\\" \\"/dev/fd/1\\"" "//1.2.3.4/share" --password p4ss'`, which does not work, because of double backslashes (it would work, at least directly in a BASH shell, with single backslashes).

`smbclient` requires the paths (`./file.txt` and `/dev/fd/1` in the example) to be either unquoted or quoted with double quotes. If I could choose, I’d use single quotes to quote whole `-c` option value and then the double quotes would not need to be escaped (i.e. `smbclient -U server -c 'get "./file.txt" "/dev/fd/1"' "//1.2.3.4/share" --password p4ss`).

Is there a way to choose the quoting behaviour and quote type (single/double quotes)?

Is there a way to work this issue around?

Thanks! :pray:

**Comment by ehmicky** [comment]
Hi @tukusejssirs,

`execa()` does not perform any logic related to quoting. It just forwards the command and its arguments to `childProcess.spawn()`, which itself forwards it to the OS. Quoting is a shell feature, so it does not apply here since you are not using any shell to execute this command.

In other words, the `smbclient` will receive exactly the arguments you specified in the `args` array, including any quotes or spaces. So: 

```js
execa('smbclient', ['-U', 'server', '-c', 'get "./file.txt" "/dev/fd/1"', '//1.2.3.4/share', '--password', 'p4ss'])
```

Forwards to:

```js
childProcess.spawn('smbclient', ['-U', 'server', '-c', 'get "./file.txt" "/dev/fd/1"', '//1.2.3.4/share', '--password', 'p4ss'])
```

Which is equivalent to running the following in a Bash shell:

```bash
smbclient -U server -c 'get "./file.txt" "/dev/fd/1"' //1.2.3.4/share --password p4ss
```

The reason why `escapedCommand` is shown as `'smbclient -U server -c "get \\"./file.txt\\" \\"/dev/fd/1\\"" "//1.2.3.4/share" --password p4ss'` is not because the double quotes is double-escaped, but because JavaScript requires significant backslashes to be escaped with another backslash. It is specific to the `escapedCommand` attribute, and does not impact how the command is executed.

In a nutshell, this problem is not related to Execa, but to child process spawning in Node.js (or actually at the OS level) in general, so I will close this issue in favor of https://github.com/eflexsystems/node-samba-client/issues/50.

It actually seems like `smbclient` is interpreting the `-c` flag correctly: the error message is showing it is looking for `/dev/fd/1`, not for `"/dev/fd/1"`, so it seems like the quotes are correctly handled by `smbclient`. I do not know `smbclient`, but it seems like the problem might have more to do with opening `/dev/fd/1` instead.

**Comment by tukusejssirs** [comment]
Thanks, @ehmicky, for you reply! :pray:

Well, `/dev/fd/1` is owned by `root` user. I ran the snippet from the OP as `root`, therefore I presume the spawned process is run as `root` as well. Could you confirm this is the case please?

**Comment by ehmicky** [comment]
Yes, this should be the case (note: this is not specific to Execa either, but to child process spawning in general).

**Comment by tukusejssirs** [comment]
You are right, it is not an issue of Execa nor `child-spawn`. All I needed to as is `{stdio: [0, 1, 2, 3]}` to options of `execa()` (which will pass it to `childProcess.spawn()`) and all works as expected ([src](https://stackoverflow.com/a/67133903/3408342)).

Thanks for bearing with me! :pray:
